pkg/versioned/map.go

```diff
@@ -0,0 +1,170 @@
+// inserted does is not mapped yet or if the object has a newer version and
+// is not deep equaled than the object already stored.
+func (m *ComparableMap) AddEqual(uuid UUID, newObj Object) bool {
+	oldObj, ok := m.Map.Get(uuid)
+	if ok {
+		// small performance optimization where we only add
+		// an object if the version is newer than the one we have.
+		if newObj.CompareVersion(oldObj) > 0 {
+			m.Map.Add(uuid, newObj)
+			return m.DeepEquals(oldObj.Data, newObj.Data)
+		}
+		return true
+	}
+	m.Map.Add(uuid, newObj)
+	return false
+}
+
+// SyncComparableMap is a thread-safe wrapper around ComparableMap.
+type SyncComparableMap struct {
+	mutex *lock.RWMutex
+	cm    *ComparableMap
+}
+
+// NewSyncComparableMap returns a thread-safe ComparableMap.
+func NewSyncComparableMap(def DeepEqualFunc) *SyncComparableMap {
+	return &SyncComparableMap{
+		mutex: &lock.RWMutex{},
+		cm:    NewComparableMap(def),
+	}
+}
+
+// Add maps the uuid to the given obj without any comparison.
+func (sm *SyncComparableMap) Add(uuid UUID, obj Object) {
+	sm.mutex.Lock()
+	sm.cm.Add(uuid, obj)
+	sm.mutex.Unlock()
+}
+
+// AddEqual maps `uuid` to `newObj` if the object to be inserted has a newer
+// Version than the one already mapped in the map. Returns false if the object
+// inserted does is not mapped yet or if the object has a newer version and
+// is not deep equaled than the object already stored.
+func (sm *SyncComparableMap) AddEqual(uuid UUID, obj Object) bool {
+	sm.mutex.Lock()
+	added := sm.cm.AddEqual(uuid, obj)
+	sm.mutex.Unlock()
+	return added
+}
+
+// Delete deletes the value that maps uuid in the map. Returns true of false
+// if the object existed in the map before deletion.
+func (sm *SyncComparableMap) Delete(uuid UUID) bool {
+	sm.mutex.Lock()
+	exists := sm.cm.Delete(uuid)
+	sm.mutex.Unlock()
+	return exists
+}
+
+// Get returns the object that maps to the given uuid and returns true or false
+// either the object exists or not.
+func (sm *SyncComparableMap) Get(uuid UUID) (Object, bool) {
+	sm.mutex.RLock()
+	v, e := sm.cm.Get(uuid)
+	sm.mutex.RUnlock()
+	return v, e
+}
+
+// DoLocked is a thread-safe function that can be used to perform multiple
+// operations in the map atomically.
+// Parameters:
+//  * iterate: if not nil, the ComparableMap will iterate over all keys and call
+//    this function for every key-value pair. Both key and value should only
+//    be used for read.
+//  * replace: if not nil, replace is called with the internal ComparableMap,
+//    the returned ComparableMap will be set as the new internal ComparableMap.
+//    If an error is returned, the replace operation won't take place.
+// In case both `iterate` and `replace` are provided, `iterate` is executed
+// first and `replace` is executed afterwards.
+func (sm *SyncComparableMap) DoLocked(iterate func(key UUID, value Object), replace func(old *ComparableMap) (*ComparableMap, error)) error {
```

Comment:
@joestringer no, it never occurred to be. While developing function was worse than it is right now. I just thought it would be better to do the 2 operations, if someone wants to have an iterate and replace func atomically.